File: test/com/way_nifty/strcat/scmj/SchemeInRubyTest.java
package com.way_nifty.strcat.scmj;

import static org.junit.Assert.*;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.StringReader;

import org.junit.Test;

public class SchemeInRubyTest {

    private String rubyFormToScheme(InputStream in) throws IOException {
        StringBuilder buf = new StringBuilder();
        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
        int ch;
        while ((ch = reader.read()) != -1) {
            switch (ch) {
            case '[':
                buf.append('(');
                break;
            case ']':
                buf.append(')');
                break;
            case ',':
                buf.append(' ');
                break;
            case '#':
                buf.append(';');
                break;
            case ':':
                break;
            default:
                buf.append((char) ch);
                break;
            }
        }
        return buf.toString();
    }
    
    private String toString(Expr expr) {
        return Printer.toString(expr);
    }

    @Test
    public void testFromRubyForm() throws Exception {
        InputStream in =
                getClass().getResourceAsStream("scheme_in_ruby_test.rb");
        String schemeSrc = rubyFormToScheme(in);
        in.close();
        
        Parser parser = new Parser(new StringReader(schemeSrc));
        Context context = new Context();
        Cell suite = (Cell) parser.readExpr();
        
        for (Cell cell : new CellIterator(suite)) {
            Cell pair = (Cell) cell.getCar();
            Expr expr = pair.nth(0);
            Expr expect = pair.nth(1);
            Expr actual = context.eval(expr);
            
            System.out.println("expr: " + toString(expr));
            System.out.println("expect: " + toString(expect));
            System.out.println("actual: " + toString(actual));
            
            assertEquals(toString(expect), toString(actual));
        }
    }

    @Test
    public void testREPL() throws Exception {
        Interpreter itp = new Interpreter();
        
        Expr result = itp.run(new StringReader(
                "(define (length list) (if (null? list) 0 (+ (length (cdr list)) 1)))"));
        assertEquals("length", toString(result));

        result = itp.run(new StringReader(
                "(length (list 1 2 3))"));
        assertEquals("3", toString(result));

        result = itp.run(new StringReader(
                "(letrec ((fact (lambda (n) (if (< n 1) 1 (* n (fact (- n 1))))))) (fact 3))"));
        assertEquals("6", toString(result));

        result = itp.run(new StringReader(
                "(let ((x 1)) (let ((dummy (set! x 2))) x))"));
        assertEquals("2", toString(result));
    }

    @Test
    public void testClosure() throws Exception {
        Interpreter itp = new Interpreter();
        
        String defineMakecounter =
                "(define (makecounter)" +
                "  (let ((count 0))" +
                "    (lambda ()" +
                "      (let ((dummy (set! count (+ count 1))))" +
                "   count))))";
        Expr result = itp.run(new StringReader(defineMakecounter));
        assertEquals("makecounter", toString(result));

        result = itp.run(new StringReader("(define inc (makecounter))"));
        assertEquals("inc", toString(result));

        result = itp.run(new StringReader("(inc)"));
        assertEquals("1", toString(result));

        result = itp.run(new StringReader("(inc)"));
        assertEquals("2", toString(result));
    }
}

File: test/com/way_nifty/strcat/scmj/scheme_in_ruby_test.rb
# scheme_in_ruby.rb (2015-05-08 version 1.0.1)からテストコード部分を抽出し一部修正。
# SchemeInRubyTest.java から読み込みれて実行される。

#$programs_expects =
  [
   # test environment
   [[[:lambda, [:x],
      [:+, 
       [[:lambda, [:x], :x], 2],
       :x]], 
     1],
    3],
   # test let
   [[:let, [[:x, 2], [:y, 3]], [:+, :x, :y]],
    5],
   [[:let, [[:x, 2] , [:y, 3]], [[:lambda, [:x, :y], [:+, :x, :y]], :x, :y]],
     5],
   [[:let, [[:add, [:lambda, [:x, :y], [:+, :x, :y]]]], [:add, 2, 3]],
    5],
   # test if
   [[:if, [:>, 3, 2], 1, 0],
    1],
   # test letrec
   [[:letrec, 
     [[:fact,
       [:lambda, [:n], [:if, [:<, :n, 1], 1, [:*, :n, [:fact, [:-, :n, 1]]]]]]], 
     [:fact, 3]], 
    6],
   # test cond 
   [[:cond, 
    [[:<, 2, 1], 0],
    [[:<, 2, 1], 1],
    [:else, 1]], 
    1],
   # test define
   [[:define, [:length, :list], 
    [:if, [:null?, :list], 0, 
     [:+, [:length, [:cdr, :list]], 1]]], 
    :length], # nil],
   [[:length, [:list, 1, 2]],
    2],
   [[:define, [:id, :x], :x],
    :id], # nil],
   [[:id, 3],
    3],
   [[:define, :x, [:lambda, [:x], :x]],
    :x], # nil],
   [[:x, 3],
    3],
   [[:define, :x, 5],
    :x], # nil],
   [:x,
    5],
   # test set!
   [[:let, [[:x, 1]],
     [:let, [[:dummy, [:set!, :x, 2]]],
      :x]], 2],
   # test list
   [[:list, 1],
    [1]],
   # test repl
#   [parse('(define (length list) (if  (null? list) 0 (+ (length (cdr list)) 1)))'),
#    nil],
#   [parse('(length (list 1 2 3))'), 
#    3],
#   [parse('(letrec ((fact (lambda (n) (if (< n 1) 1 (* n (fact (- n 1))))))) (fact 3))'),
#    6],
#   [parse('(let ((x 1)) (let ((dummy (set! x 2))) x))'),
#    2],
   # test fixed point
   # fact(0) = 1
   [[:let, 
     [[:fact,
       [:lambda, [:n], [:if, [:<, :n, 1], 1, [:*, :n, [:fact, [:-, :n, 1]]]]]]], 
     [:fact, 0]], 1],
   # fact(1) = 1
   [[:let, 
     [[:fact,
       [:lambda, [:n], 
        [:if, [:<, :n, 1], 1, 
         [:*, :n, 
          [:let, 
           [[:fact,
             [:lambda, [:n], [:if, [:<, :n, 1], 1, [:*, :n, [:fact, [:-, :n, 1]]]]]]],
           [:fact, [:-, :n, 1]]]]]]]], 
     [:fact, 1]], 1],
   # fact(2) = 2
   [[:let, 
     [[:fact,
       [:lambda, [:n], 
        [:if, [:<, :n, 1], 1, 
         [:*, :n, 
          [:let, 
           [[:fact,
             [:lambda, [:n], [:if, [:<, :n, 1], 1, [:*, :n, [:fact, [:-, :n, 1]]]]]]],
           [:let, 
            [[:fact,
              [:lambda, [:n], [:if, [:<, :n, 1], 1, [:*, :n, [:fact, [:-, :n, 1]]]]]]],
            
            [:fact, [:-, :n, 1]]]]]]]]], 
     [:fact, 2]], 2],
   # closure
#   [parse(
#<<EOS
#(define (makecounter)
#  (let ((count 0))
#    (lambda ()
#      (let ((dummy (set! count (+ count 1))))
#	count))))
#EOS
#), nil],
#   [parse('(define inc (makecounter))'), nil],
#   [parse('(inc)'), 1],
#   [parse('(inc)'), 2],
  ]
